test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py

```diff
@@ -208,221 +138,115 @@ def __new__(cls, class_name, parents, attrs):
 
-    def serialize(self):
-        from .ssz_impl import serialize
-        return serialize(self, self.__class__)
+    def __setitem__(self, k, v):
+        self.items[k] = v
 
-    def hash_tree_root(self):
-        from .ssz_impl import hash_tree_root
-        return hash_tree_root(self, self.__class__)
+    def __len__(self):
+        return len(self.items)
 
     def __repr__(self):
-        return repr({'length': self.__class__.length, 'items': self.items})
-
-    def __getitem__(self, key):
-        return self.items[key]
-
-    def __setitem__(self, key, value):
-        self.items[key] = value
+        return repr(self.items)
 
     def __iter__(self):
         return iter(self.items)
 
-    def __len__(self):
-        return len(self.items)
-
     def __eq__(self, other):
-        return self.hash_tree_root() == other.hash_tree_root()
+        return self.items == other.items
 
+class List(AbstractList, metaclass=AbstractListMeta):
+    def value_check(self, value):
+        return len(value) <= self.__class__.length and super().value_check(value)
 
-# SSZ BytesN
-# -----------------------------
-
-
-def _is_bytes_n_instance_of(a, b):
-    # Other has to be a Bytes derivative class to be a BytesN
-    if not issubclass(b, bytes):
-        return False
-    elif not hasattr(b, 'length'):
-        # BytesN (b) is not an instance of BytesN[X] (a)
-        return False
-    elif not hasattr(a, 'length'):
-        # BytesN[X] (b) is an instance of BytesN (a)
-        return True
-    else:
-        # BytesN[X] (a) is an instance of BytesN[X] (b)
-        return a.length == b.length
-
-
-def _is_equal_bytes_n_type(a, b):
-    # Other has to be a Bytes derivative class to be a BytesN
-    if not issubclass(b, bytes):
-        return False
-    elif not hasattr(a, 'length'):
-        if not hasattr(b, 'length'):
-            # BytesN == BytesN
-            return True
-        else:
-            # BytesN != BytesN[X]
-            return False
-    elif not hasattr(b, 'length'):
-        # BytesN[X] != BytesN
-        return False
-    else:
-        # BytesN[X] == BytesN[X]
-        return a.length == b.length
-
-
-class BytesNMeta(type):
-    def __new__(cls, class_name, parents, attrs):
-        out = type.__new__(cls, class_name, parents, attrs)
-        if 'length' in attrs:
-            setattr(out, 'length', attrs['length'])
-        out._name = 'BytesN'
-        out.elem_type = byte
-        return out
+    def default(self):
+        return []
 
-    def __getitem__(self, n):
-        return self.__class__(self.__name__, (BytesN,), {'length': n})
+class Vector(AbstractList, metaclass=AbstractListMeta):
+    def value_check(self, value):
+        return len(value) == self.__class__.length and super().value_check(value)
 
-    def __subclasscheck__(self, sub):
-        return _is_bytes_n_instance_of(self, sub)
+    def default(self):
+        return [get_zero_value(self.__class__.elem_type) for _ in range(self.__class__.length)]
```

Comment:
you are removing `get_zero_value` right? Shouldn't we call something like `self.__class__.elem_type.default()` instead now?

Reply:
I suppose we could. I was originally designing the code so that we could put values at the bottom that are not SSZ types, but if that's not an objective we can definitely get rid of `get_zero_value` and just use `typ.default()`